opm/simulators/linalg/bda/rocsparseSolverBackend.cpp

```diff
@@ -0,0 +1,612 @@
+/*
+  Copyright 2023 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <sstream>
+#include <stdexcept>
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/ErrorMacros.hpp>
+#include <dune/common/timer.hh>
+
+// WellContributions are included via the solver
+// MultisegmentWellContribution includes the cuda runtime if found by CMake
+// this leads to inclusion of both amd_hip_vector_types.h and vector_types.h
+// which both define vector types like uchar2, short3 and double4.
+// Restore the value (if defined) afterwards.
+#ifdef HAVE_CUDA
+#define HIP_HAVE_CUDA_DEFINED HAVE_CUDA
+#endif
+
+#undef HAVE_CUDA
+
+#include <opm/simulators/linalg/bda/rocsparseSolverBackend.hpp>
+
+#include <opm/simulators/linalg/bda/BdaResult.hpp>
+
+#include <hip/hip_runtime_api.h>
+#include <hip/hip_version.h>
+
+#ifdef HIP_HAVE_CUDA_DEFINED HAVE_CUDA
+#define HAVE_CUDA HIP_HAVE_CUDA_DEFINED
+#undef HIP_HAVE_CUDA_DEFINED HAVE_CUDA
```

Comment:
`HAVE_CUDA` doesn't do anything here, and will just emit a warning.  What's the intended semantics of this directive?